src/Catel.MVVM/Catel.MVVM.Shared/Collections/FastObservableCollection.cs

```diff
@@ -326,6 +422,112 @@ protected override void OnPropertyChanged(PropertyChangedEventArgs e)
+            }
+
+            // Call base
+            base.InsertItem(index, item);
+
+            if (_suspensionContext != null && _suspensionContext.Mode == SuspensionMode.Adding)
+            {
+                // Remember
+                _suspensionContext.NewItems.Add(item);
+                _suspensionContext.NewItemIndices.Add(index);
+            }
+        }
+
+        /// <summary>
+        /// Moves the item at the specified index to a new location in the collection.
+        /// </summary>
+        /// <param name="oldIndex">The zero-based index specifying the location of the item to be moved.</param><param name="newIndex">The zero-based index specifying the new location of the item.</param>
+        protected override void MoveItem(int oldIndex, int newIndex)
+        {
+            // Check
+            if (_suspensionContext != null && _suspensionContext.Mode != SuspensionMode.None)
+            {
+                throw Log.ErrorAndCreateException<InvalidOperationException>("Moving items is not allowed in modes other than mixed.");
+            }
+
+            // Call base
+            base.MoveItem(oldIndex, newIndex);
+        }
+
+        /// <summary>
+        /// Removes the item at the specified index of the collection.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element to remove.</param>
+        protected override void RemoveItem(int index)
+        {
+            // Check
+            if (_suspensionContext != null && _suspensionContext.Mode == SuspensionMode.Adding)
+            {
+                throw Log.ErrorAndCreateException<InvalidOperationException>("Removing items is not allowed in adding mode.");
+            }
+
+            // Get item
+            T item;
+            if (_suspensionContext != null && _suspensionContext.Mode == SuspensionMode.Removing)
+            {
+                item = this[index];
+            }
+            else
+            {
+                item = default(T);
+            }
+
+            // Call base
+            base.RemoveItem(index);
+
+            if (_suspensionContext != null && _suspensionContext.Mode == SuspensionMode.Removing)
+            {
+                // Remember
+                _suspensionContext.OldItems.Add(item);
+                _suspensionContext.OldItemIndices.Add(index);
+            }
+        }
+
+        /// <summary>
+        /// Replaces the element at the specified index.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element to replace.</param><param name="item">The new value for the element at the specified index.</param>
+        protected override void SetItem(int index, T item)
+        {
+            // Check
+            if (_suspensionContext != null && _suspensionContext.Mode != SuspensionMode.None)
+            {
+                throw Log.ErrorAndCreateException<InvalidOperationException>("Replacing items is not allowed in modes other than mixed.");
```

Comment:
I think I understand why you are throwing the exception, but I think the text should be a bit different. Something like:

```
$"Replacing items is only allowed in SuspensionMode.None, current mode is '{_suspensionContext.Mode}'"
```